### PR TITLE
feat: Step divider manipulation

### PR DESCRIPTION
**Pull request by justinkambic**
## Summary

**Note to reviewers**: do not attempt to review this until #137 is merged.

Depends on #137.

Related to #64.

Adds the API required to manipulate steps for the insertion/merging/rearrangement of recorded steps.

## Implementation details

This patch adds several new handlers to the `StepsContext` which provide the functionality to manipulate the steps list in new and interesting ways. These changes allow the user to introduce a new step between two existing steps. They will also allow the user to rearrange two steps by index, per the linked issue.

These changes would also introduce a new `merge` function to remove a step without deleting the nested actions. If the user wants to remove a step divider but avoid breaking their journey, we can provide this feature by merging the target step with the previous one.

## How to validate this change

I have added numerous unit tests to cover the new functionality. I haven't implemented the UI piece of this yet; if reviewers prefer I can update this PR to have the UI changes as well, or we can do them in a follow-up.

### CODE REVIEW

**Review by lucasfcosta — APPROVED** [review]
LGTM (only non-blocking comments)! Thanks for this PR 🙌 

I actually really like that we have split the UI changes from the API changes here to keep the PR small.

I'm open to both adding them to this PR or merging this and opening a separate PR later, although I prefer the latter.

I also ran this E2E just to be sure the existing UI didn't break.

File: src/hooks/useStepsContext.ts [hunk 1]
```diff
@@ -65,6 +65,47 @@ export function useStepsContext(): IStepsContext {
...
+        oldSteps[indexA] = oldSteps[indexB];
+        oldSteps[indexB] = placeholder;
+        return oldSteps;
+      });
+    },
+    onSplitStep: (stepIndex, actionIndex) => {
+      if (actionIndex === 0)
+        throw Error(
+          `Split procedure received action index ${actionIndex}. Cannot remove all actions from a step.`
+        );
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/hooks/useStepsContext.ts:90]: Sure, this is an example of copy written by me in the heat of the moment and then never updated.

[the remark this answers is not in the document]
Comment (justinkambic) [reply src/hooks/useStepsContext.ts]: https://github.com/elastic/synthetics-recorder/pull/148/commits/7a0ecfdf8862054779449d03b8c1da8cfb8bf65c

File: src/hooks/useStepsContext.ts [hunk 2]
```diff
@@ -65,6 +65,47 @@ export function useStepsContext(): IStepsContext {
...
+          ...steps[indexToInsert],
+          ...steps[indexToRemove],
+        ];
+        oldSteps.splice(indexToRemove, 1);
+        return oldSteps;
+      });
+    },
+    onRearrangeSteps: (indexA, indexB) => {
+      setSteps(oldSteps => {
+        const placeholder = [...steps[indexA]];
```
[the remark this answers is not in the document]
Comment (justinkambic) [reply src/hooks/useStepsContext.ts:80]: I think the reason I wrote it this way was from a desire to create a new array, vs. referencing the existing one.

Redux PTSD.

[the remark this answers is not in the document]
Comment (justinkambic) [reply src/hooks/useStepsContext.ts]: https://github.com/elastic/synthetics-recorder/pull/148/commits/c4a596875925ae8f3bb870d9a24b217546fbb851

**Comment by justinkambic** [comment]
> I actually really like that we have split the UI changes from the API changes here to keep the PR small.

I think it's more focused as well. Given that there's no user impact by adding this code, I think it'll be fine and cleaner to merge this and then follow-up with a UX enhancement PR.

**Pull request merged**